File: examples/busywaiting/flexiblespecs/spinlock/spinlock.c
// Justus Fasse and Bart Jacobs. Expressive modular verification of termination for busy-waiting programs. 2023.

#include <stdlib.h>
#include "atomics.h"
#include "spinlock.h"

struct spinlock {
    int locked;
    //@ predicate(bool) inv_;
};

/*@

predicate_ctor spinlock_inv(spinlock_t spinlock, predicate(bool) inv)() =
    spinlock->locked |-> ?locked &*& inv(locked != 0);

predicate spinlock(spinlock_t spinlock; predicate(bool) inv) =
    spinlock->inv_ |-> inv &*&
    malloc_block_spinlock(spinlock) &*&
    atomic_space(create_spinlock, spinlock_inv(spinlock, inv));

@*/

spinlock_t create_spinlock()
//@ requires exists<predicate(bool)>(?inv) &*& inv(false);
//@ ensures spinlock(result, inv);
//@ terminates;
{
  spinlock_t result = malloc(sizeof(struct spinlock));
  if (result == 0) abort();
  result->locked = 0;
  //@ result->inv_ = inv;
  //@ close spinlock_inv(result, inv)();
  //@ create_atomic_space(create_spinlock, spinlock_inv(result, inv));
  return result;
}

void spinlock_acquire(spinlock_t spinlock)
//@ requires [?f]spinlock(spinlock, ?inv) &*& is_spinlock_acquire_ghost_op(?ghop, inv, ?pre, ?post, currentThread) &*& pre();
//@ ensures [f]spinlock(spinlock, inv) &*& post();
//@ terminates;
{
  for (;;)
  //@ invariant [f]spinlock(spinlock, inv) &*& is_spinlock_acquire_ghost_op(ghop, inv, pre, post, currentThread) &*& pre();
  {
    int result;
    //@ open spinlock(spinlock, inv);
    {
      /*@
      predicate pre_() = is_spinlock_acquire_ghost_op(ghop, inv, pre, post, currentThread) &*& pre();
      predicate post_(int result_) =
        result_ == 0 ?
          post()
        :
          is_spinlock_acquire_ghost_op(ghop, inv, pre, post, currentThread) &*& pre() &*& call_perm_(currentThread, spinlock_acquire);
      @*/
      /*@
      produce_lemma_function_pointer_chunk compare_and_swap_int_ghost_op(create_spinlock, spinlock_inv(spinlock, inv), &spinlock->locked, 0, 1, pre_, post_, currentThread)() {
        assert is_compare_and_swap_int_op(?op, _, _, _, _, _);
        open spinlock_inv(spinlock, inv)();
        open pre_();
        assert spinlock->locked |-> ?locked;
        op();
        ghop();
        if (locked == 0) {
          leak is_spinlock_acquire_ghost_op(_, _, _, _, currentThread);
        } else {
        }
        close spinlock_inv(spinlock, inv)();
        close post_(locked);
      };
      @*/
      //@ close pre_();
      result = compare_and_swap_int(&spinlock->locked, 0, 1);
      //@ open post_(result);
    }
    if (result == 0) {
      break;
    }
  }
}

void spinlock_release(spinlock_t spinlock)
//@ requires [?f]spinlock(spinlock, ?inv) &*& is_spinlock_release_ghost_op(?ghop, inv, ?pre, ?post) &*& pre();
//@ ensures [f]spinlock(spinlock, inv) &*& post();
//@ terminates;
{
  //@ open spinlock(spinlock, inv);
  {
    /*@
    predicate pre_() = is_spinlock_release_ghost_op(ghop, inv, pre, post) &*& pre();
    predicate post_() = post();
    @*/
    /*@
    produce_lemma_function_pointer_chunk atomic_store_int_ghost_op(create_spinlock, spinlock_inv(spinlock, inv), &spinlock->locked, 0, pre_, post_, currentThread)() {
      assert is_atomic_store_int_op(?op, _, _, _, _);
      open spinlock_inv(spinlock, inv)();
      open pre_();
      op();
      ghop();
      leak is_spinlock_release_ghost_op(_, _, _, _);
      close spinlock_inv(spinlock, inv)();
      close post_();
    };
    @*/
    //@ close pre_();
    atomic_store_int(&spinlock->locked, 0);
    //@ leak is_atomic_store_int_ghost_op(_, _, _, _, _, _, _, _);
    //@ open post_();
  }
  //@ close [f]spinlock(spinlock, inv);
}

void spinlock_dispose(spinlock_t spinlock)
//@ requires spinlock(spinlock, ?inv);
//@ ensures inv(_);
//@ terminates;
{
    //@ open spinlock(spinlock, _);
    //@ destroy_atomic_space();
    //@ open spinlock_inv(spinlock, inv)();
    free(spinlock);
}

File: examples/busywaiting/flexiblespecs/ticketlock-java/Spinlock.java
// tab_size:2

// Justus Fasse and Bart Jacobs. Expressive modular verification of termination for busy-waiting programs. 2023.

public final class Spinlock {

  //@ list<int> ns;
  //@ level level;
  private AtomicLong lk;

  //@ predicate valid(list<int> ns, level level) = this.ns |-> ns &*& this.level |-> level &*& [_]lk |-> ?_lk &*& _lk != null &*& 0 <= level_subspace_nb_dims(level);

  //@ predicate held() = [_]lk |-> ?_lk &*& [1/2]_lk.state(true);
  
  //@ predicate state(boolean held) = [_]lk |-> ?_lk &*& [1/2]_lk.state(?value) &*& value == 0 ? [1/2]_lk.state(false) &*& held == false : value == 1 &*& held == true;

  public Spinlock()
  //@ requires exists<pair<list<int>, level> >(pair(?ns, ?level)) &*& Spinlock_level_nb_dims <= level_subspace_nb_dims(level);
  //@ ensures [_]valid(ns, level) &*& state(false);
  //@ terminates;
  {
    //@ this.ns = ns;
    //@ this.level = level;
    lk = new AtomicLong();
    //@ close state(false);
  }

  public void acquire()
  /*@
  requires
    [_]valid(?ns, ?level) &*&
    obs(currentThread, ?p, ?obs) &*& forall(map(snd, obs), (level_subspace_lt)(level)) == true &*&
    is_Spinlock_sep(?sep, this, ns, ?waitInv, ?R) &*&
    is_Spinlock_wait_ghost_op(?wop, this, R, currentThread, p, level, waitInv) &*&
    is_Spinlock_acquire_ghost_op(?aop, this, R, p, obs, ?post, currentThread) &*&
    waitInv();
  @*/
  //@ ensures held() &*& post();
  //@ terminates;
  {
    //@ AtomicLong _lk = lk;
    for (;;)
    /*@
    invariant
      [_]lk |-> _lk &*&
      obs(currentThread, p, obs) &*&
      is_Spinlock_sep(sep, this, ns, waitInv, R) &*&
      is_Spinlock_wait_ghost_op(wop, this, R, currentThread, p, level, waitInv) &*&
      is_Spinlock_acquire_ghost_op(aop, this, R, p, obs, post, currentThread) &*&
      waitInv();
    @*/
    {
      boolean success;
      {
        /*@
        predicate pre_() =
          [_]lk |-> _lk &*&
          obs(currentThread, p, obs) &*&
          is_Spinlock_sep(sep, this, ns, waitInv, R) &*&
          is_Spinlock_wait_ghost_op(wop, this, R, currentThread, p, level, waitInv) &*&
          is_Spinlock_acquire_ghost_op(aop, this, R, p, obs, post, currentThread) &*&
          waitInv();
        predicate post_(boolean success_) =
          is_Spinlock_sep(sep, this, ns, waitInv, R) &*&
          is_Spinlock_wait_ghost_op(wop, this, R, currentThread, p, level, waitInv) &*&
          is_Spinlock_acquire_ghost_op(aop, this, R, p, obs, post, currentThread) &*&
          success_ ?
            post() &*& [1/2]_lk.state(true)
          :
            obs(currentThread, p, obs) &*&
            waitInv() &*&
            call_perm_(currentThread, Spinlock.class);
        @*/
        /*@
        produce_lemma_function_pointer_chunk AtomicLong_compareAndSet_ghost_op(_lk, 0, 1, pre_, post_)(op) {
          open pre_();
          sep();
          open state(?held);
          assert [_]_lk.state(?value);
          op();
          if (held) {
            if (!forall(map(snd, obs), (level_lt)(level))) {
              level badLevel = not_forall(map(snd, obs), (level_lt)(level));
              forall_elim(map(snd, obs), (level_subspace_lt)(level), badLevel);
              level_subspace_lt_level_lt(level, {}, badLevel);
              assert false;
            }
            wop();
            close post_(false);
          } else {
            aop();
            close post_(true);
          }
        };
        @*/
        //@ close pre_();
        success = lk.compareAndSet(0, 1);
        //@ open post_(success);
      }
      if (success)
        break;
    }
    //@ close held();
  }

  public void release()
  /*@
  requires
    [_]valid(?ns, ?level) &*&
    held() &*&
    is_Spinlock_sep(?sep, this, ns, ?pre, ?R) &*&
    is_Spinlock_release_ghost_op(?ghop, this, R, ?post, currentThread) &*& pre();
  @*/
  //@ ensures post(?p, ?obs) &*& obs(currentThread, p, obs);
  //@ terminates;
  {
    //@ AtomicLong _lk = lk;
    {
      /*@
      predicate pre_() =
        [_]lk |-> _lk &*& [1/2]_lk.state(true) &*&
        is_Spinlock_sep(sep, this, ns, pre, R) &*&
        is_Spinlock_release_ghost_op(ghop, this, R, post, currentThread) &*&
        pre();
      predicate post_() =
        post(?p, ?obs) &*& obs(currentThread, p, obs);
      @*/
      /*@
      produce_lemma_function_pointer_chunk AtomicLong_set_ghost_op(_lk, 0, pre_, post_)(op) {
        open pre_();
        sep();
        open state(?held);
        op();
        ghop();
        close post_();
      };
      @*/
      //@ close pre_();
      lk.set(0);
      //@ open post_();
    }
  }

}
